// @cubejs-backend/schema-compiler/dist/src/scaffolding/ScaffoldingSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScaffoldingSchema = exports.MemberType = void 0;
const inflection_1 = __importDefault(require("inflection"));
const ramda_1 = __importDefault(require("ramda"));
const compiler_1 = require("../compiler");
var ColumnType;
(function (ColumnType) {
    ColumnType["Time"] = "time";
    ColumnType["Number"] = "number";
    ColumnType["String"] = "string";
})(ColumnType || (ColumnType = {}));
var MemberType;
(function (MemberType) {
    MemberType["Measure"] = "measure";
    MemberType["Dimension"] = "dimension";
    MemberType["None"] = "none";
})(MemberType = exports.MemberType || (exports.MemberType = {}));
const MEASURE_DICTIONARY = [
    'amount',
    'price',
    'count',
    'balance',
    'total',
    'number',
    'cost',
    'qty',
    'quantity',
    'duration',
    'value'
];
const DRILL_MEMBERS_DICTIONARY = [
    'id',
    'name',
    'title',
    'firstname',
    'first_name',
    'lastname',
    'last_name',
    'createdat',
    'created_at',
    'created',
    'timestamp',
    'city',
    'country',
    'date'
];
const idRegex = '_id$|id$';
class ScaffoldingSchema {
    constructor(dbSchema, options = {}) {
        this.dbSchema = dbSchema;
        this.options = options;
    }
    cubeDescriptors(tableNames) {
        const cubes = this.generateForTables(tableNames);
        function member(type) {
            return (value) => ({
                memberType: type,
                ...ramda_1.default.pick(['name', 'title', 'types', 'isPrimaryKey', 'included', 'isId'], value)
            });
        }
        return cubes.map((cube) => ({
            cube: cube.cube,
            tableName: cube.tableName,
            table: cube.table,
            schema: cube.schema,
            members: (cube.measures || []).map(member(MemberType.Measure))
                .concat((cube.dimensions || []).map(member(MemberType.Dimension))),
            joins: cube.joins
        }));
    }
    generateForTables(tableNames) {
        this.prepareTableNamesToTables(tableNames);
        return tableNames.map(tableName => this.tableSchema(tableName, true));
    }
    prepareTableNamesToTables(tableNames) {
        this.tableNamesToTables = ramda_1.default.pipe(
        // @ts-ignore
        ramda_1.default.unnest, ramda_1.default.groupBy(n => n[0]), ramda_1.default.map(groupedNameToDef => groupedNameToDef.map(nameToDef => nameToDef[1])))(
        // @ts-ignore
        tableNames.map(tableName => {
            const [schema, table] = this.parseTableName(tableName);
            const tableDefinition = this.resolveTableDefinition(tableName);
            const definition = {
                schema, table, tableDefinition, tableName
            };
            const tableizeName = inflection_1.default.tableize(table);
            const parts = tableizeName.split('_');
            const tableNamesFromParts = ramda_1.default.range(0, parts.length - 1).map(toDrop => inflection_1.default.tableize(ramda_1.default.drop(toDrop, parts).join('_')));
            const names = ramda_1.default.uniq([table, tableizeName].concat(tableNamesFromParts));
            return names.map(n => [n, definition]);
        }));
    }
    resolveTableDefinition(tableName) {
        const [schema, table] = this.parseTableName(tableName);
        if (!this.dbSchema[schema]) {
            throw new compiler_1.UserError(`Can't resolve ${tableName}: '${schema}' does not exist`);
        }
        if (!this.dbSchema[schema][table]) {
            throw new compiler_1.UserError(`Can't resolve ${tableName}: '${table}' does not exist`);
        }
        return this.dbSchema[schema][table];
    }
    tableSchema(tableName, includeJoins) {
        const [schema, table] = this.parseTableName(tableName);
        const tableDefinition = this.resolveTableDefinition(tableName);
        const dimensions = this.dimensions(tableDefinition);
        return {
            cube: inflection_1.default.camelize(table),
            tableName,
            schema,
            table,
            measures: this.numberMeasures(tableDefinition),
            dimensions,
            drillMembers: this.drillMembers(dimensions),
            joins: includeJoins ? this.joins(tableName, tableDefinition) : []
        };
    }
    parseTableName(tableName) {
        let schemaAndTable;
        if (Array.isArray(tableName)) {
            schemaAndTable = tableName;
        }
        else {
            schemaAndTable = tableName.match(/(["`].*?["`]|[^`".]+)+(?=\s*|\s*$)/g);
        }
        if (schemaAndTable.length !== 2) {
            throw new compiler_1.UserError(`Incorrect format for '${tableName}'. Should be in '<schema>.<table>' format`);
        }
        return schemaAndTable;
    }
    dimensions(tableDefinition) {
        return this.dimensionColumns(tableDefinition).map(column => {
            var _a;
            const res = {
                name: column.name,
                types: [column.columnType || this.columnType(column)],
                title: inflection_1.default.titleize(column.name),
            };
            if (column.columnType !== 'time') {
                res.isPrimaryKey = ((_a = column.attributes) === null || _a === void 0 ? void 0 : _a.includes('primaryKey')) ||
                    column.name.toLowerCase() === 'id';
            }
            return res;
        });
    }
    numberMeasures(tableDefinition) {
        return tableDefinition.filter(column => (!column.name.startsWith('_') &&
            (this.columnType(column) === 'number') &&
            (this.options.includeNonDictionaryMeasures ? column.name.toLowerCase() !== 'id' : this.fromMeasureDictionary(column)))).map(column => ({
            name: column.name,
            types: ['sum', 'avg', 'min', 'max'],
            title: inflection_1.default.titleize(column.name),
            ...(this.options.includeNonDictionaryMeasures ? { included: this.fromMeasureDictionary(column) } : null)
        }));
    }
    fromMeasureDictionary(column) {
        return !column.name.match(new RegExp(idRegex, 'i')) && !!MEASURE_DICTIONARY.find(word => column.name.toLowerCase().endsWith(word));
    }
    dimensionColumns(tableDefinition) {
        const dimensionColumns = tableDefinition.filter(column => {
            var _a;
            return !column.name.startsWith('_') && this.columnType(column) === 'string' || ((_a = column.attributes) === null || _a === void 0 ? void 0 : _a.includes('primaryKey')) ||
                column.name.toLowerCase() === 'id';
        });
        const timeColumns = ramda_1.default.pipe(
        // @ts-ignore
        ramda_1.default.filter(column => !column.name.startsWith('_') && this.columnType(column) === 'time'), ramda_1.default.sortBy(column => this.timeColumnIndex(column)), 
        // @ts-ignore
        ramda_1.default.map(column => ({ ...column, columnType: 'time' })) // TODO do we need it?
        // @ts-ignore
        )(tableDefinition);
        return dimensionColumns.concat(timeColumns);
    }
    joins(tableName, tableDefinition) {
        return ramda_1.default.unnest(tableDefinition
            .filter(column => (column.name.match(new RegExp(idRegex, 'i')) && column.name.toLowerCase() !== 'id'))
            .map(column => {
            const withoutId = column.name.replace(new RegExp(idRegex, 'i'), '');
            const tablesToJoin = this.tableNamesToTables[withoutId] ||
                this.tableNamesToTables[inflection_1.default.tableize(withoutId)];
            if (!tablesToJoin) {
                return null;
            }
            const columnsToJoin = tablesToJoin.map(definition => {
                if (tableName === definition.tableName) {
                    return null;
                }
                let columnForJoin = definition.tableDefinition.find(c => c.name.toLowerCase() === column.name.toLowerCase());
                columnForJoin = columnForJoin || definition.tableDefinition.find(c => c.name.toLowerCase() === 'id');
                if (!columnForJoin) {
                    return null;
                }
                return {
                    cubeToJoin: inflection_1.default.camelize(definition.table),
                    columnToJoin: columnForJoin.name,
                    tableName: definition.tableName
                };
            }).filter(ramda_1.default.identity);
            return columnsToJoin.map(columnToJoin => ({
                thisTableColumn: column.name,
                tableName: columnToJoin.tableName,
                cubeToJoin: columnToJoin.cubeToJoin,
                columnToJoin: columnToJoin.columnToJoin,
                relationship: 'belongsTo'
            }));
        })
            .filter(ramda_1.default.identity));
    }
    drillMembers(dimensions) {
        return dimensions.filter(d => this.fromDrillMembersDictionary(d));
    }
    fromDrillMembersDictionary(dimension) {
        return !!DRILL_MEMBERS_DICTIONARY.find(word => dimension.name.toLowerCase().includes(word));
    }
    timeColumnIndex(column) {
        const name = column.name.toLowerCase();
        if (name.indexOf('create') !== -1) {
            return 0;
        }
        else if (name.indexOf('update') !== -1) {
            return 1;
        }
        else {
            return 2;
        }
    }
    columnType(column) {
        const type = column.type.toLowerCase();
        if (['time', 'date'].find(t => type.indexOf(t) !== -1)) {
            return ColumnType.Time;
        }
        else if (['int', 'dec', 'double', 'numb'].find(t => type.indexOf(t) !== -1)) {
            // enums are not Numbers
            return ColumnType.Number;
        }
        else {
            return ColumnType.String;
        }
    }
}
exports.ScaffoldingSchema = ScaffoldingSchema;
